crates/neon/src/types_impl/extract/types.rs

```diff
@@ -0,0 +1,356 @@
+        extract::{private, TryFromJs},
+        private::ValueInternal,
+        JsArrayBuffer, JsBoolean, JsBuffer, JsNumber, JsString, JsTypedArray, JsValue, Value,
+    },
+};
+
+#[cfg(feature = "napi-5")]
+use crate::types::JsDate;
+
+/// Error returned when a JavaScript value is not the type expected
+pub struct TypeExpected<T: Value>(PhantomData<T>);
+
+impl<T: Value> TypeExpected<T> {
+    fn new() -> Self {
+        Self(PhantomData)
+    }
+}
+
+impl<T: Value> fmt::Display for TypeExpected<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "expected {}", T::name())
+    }
+}
+
+impl<T: Value> fmt::Debug for TypeExpected<T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("TypeExpected").field(&T::name()).finish()
+    }
+}
+
+impl<T: Value> error::Error for TypeExpected<T> {}
+
+impl<T, U: Value> ResultExt<T> for Result<T, TypeExpected<U>> {
+    fn or_throw<'a, C: Context<'a>>(self, cx: &mut C) -> NeonResult<T> {
+        match self {
+            Ok(v) => Ok(v),
+            Err(_) => cx.throw_type_error(format!("expected {}", U::name())),
+        }
+    }
+}
+
+macro_rules! from_js {
+    () => {
+        fn from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Self>
+        where
+            C: Context<'cx>,
+        {
+            Self::try_from_js(cx, v)?.or_throw(cx)
+        }
+    };
+}
+
+impl<'cx, V> TryFromJs<'cx> for Handle<'cx, V>
+where
+    V: Value,
+{
+    type Error = TypeExpected<V>;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        Ok(v.downcast(cx).map_err(|_| TypeExpected::new()))
+    }
+
+    from_js!();
+}
+
+impl<'cx, V: Value> private::Sealed for Handle<'cx, V> {}
+
+impl<'cx, T> TryFromJs<'cx> for Option<T>
+where
+    T: TryFromJs<'cx>,
+{
+    type Error = T::Error;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        if is_null_or_undefined(cx, v)? {
+            return Ok(Ok(None));
+        }
+
+        T::try_from_js(cx, v).map(|v| v.map(Some))
+    }
+
+    fn from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Self>
+    where
+        C: Context<'cx>,
+    {
+        if is_null_or_undefined(cx, v)? {
+            return Ok(None);
+        }
+
+        T::from_js(cx, v).map(Some)
+    }
+}
+
+impl<'cx, T> private::Sealed for Option<T> where T: TryFromJs<'cx> {}
+
+impl<'cx> TryFromJs<'cx> for f64 {
+    type Error = TypeExpected<JsNumber>;
+
+    fn try_from_js<C>(cx: &mut C, v: Handle<'cx, JsValue>) -> NeonResult<Result<Self, Self::Error>>
+    where
+        C: Context<'cx>,
+    {
+        let mut n = 0f64;
+
+        unsafe {
+            match sys::get_value_double(cx.env().to_raw(), v.to_local(), &mut n) {
+                sys::Status::NumberExpected => return Ok(Err(TypeExpected::new())),
```

Comment:
I think we've typically kept this kind of logic out of Neon proper and relegated to the runtime crate, but maybe I'm mis-remembering or you want to intentionally drop that convention. (Come to think of it, I don't know if that has a performance cost due to compiler limitations for cross-crate inlining.)

Reply:
I added a comment explaining the rationale and also created a follow-up issue to clean-up module orgnaization.

https://github.com/neon-bindings/neon/issues/1027